File: spec/thermostatSpec.js
'use strict';

describe('Thermostat', function(){
    var thermostat;
    beforeEach(function(){
        thermostat = new Thermostat;
    });

    describe('Starts at 20', function(){
        it ('thermostat has default temperature of 20', function(){
            expect(thermostat.getCurrentTemp()).toEqual(20);
        });
    });

    describe('Changing temp', function(){
        it ('increases the temperature', function(){
            thermostat.up();
            expect(thermostat.getCurrentTemp()).toEqual(21);
        });

        it ('decreases the temperature', function(){
            thermostat.down();
            expect(thermostat.getCurrentTemp()).toEqual(19);
        });
    });

    describe('Minimum temp', function(){
        it ('Prevents temp from going below 10', function(){
            for (var i=0; i<20; i++){    
                thermostat.down();
            }
            expect(thermostat.getCurrentTemp()).toEqual(10);
        });
    });

    describe('Power saving mode', function(){
        it('is on by default', function(){
            expect(thermostat.isPowerSavingOn()).toEqual(true);
        });

        it('can switch PSM off', function() {
            thermostat.switchPowerSavingOff();
            expect(thermostat.isPowerSavingOn()).toEqual(false);
        });

        it('can switch PSM back on', function() {
            thermostat.switchPowerSavingOff();
            expect(thermostat.isPowerSavingOn()).toEqual(false);
            thermostat.switchPowerSavingOn();
            expect(thermostat.isPowerSavingOn()).toEqual(true);
        });


        it('Sets maximum temp to 25 when on', function(){
            for(var i=0; i<10; i++){
                thermostat.up();
            }
            expect(thermostat.getCurrentTemp()).toEqual(25);
        });

        it('Sets maximum temp to 32 when off', function(){
            thermostat.switchPowerSavingOff();

            for(var i=0; i<15; i++){
                thermostat.up();
            }
            expect(thermostat.getCurrentTemp()).toEqual(32);
        });

        it('Limits temp back to 25 when power saving turned back on', function(){
            thermostat.switchPowerSavingOff();

            for(var i=0; i<15; i++){
                thermostat.up();
            }

            thermostat.switchPowerSavingOn();
            expect(thermostat.getCurrentTemp()).toEqual(25);
        })
    });

    describe('Reset temperature', function(){
        it ('Resets the temperature to 20', function(){
            thermostat.up();
            thermostat.reset();

            expect(thermostat.getCurrentTemp()).toEqual(20);
        });
    });

    describe('Energy usage', function(){
        it ('returns low-usage if temp less than 18', function(){

            for(var i=0; i<5; i++){
                thermostat.down();
            }
            expect(thermostat.getEnergyUsage()).toEqual("low-usage");
        });

        it ('returns medium-usage if temp less than or equal to 25', function(){
            expect(thermostat.getEnergyUsage()).toEqual("medium-usage");
        });

        it ('returns high-usage if temp greater than 25', function(){
            thermostat.powerSaving= false;

            for(var i=0; i<10; i++){
                thermostat.up();
            }

            expect(thermostat.getEnergyUsage()).toEqual("high-usage");
        });
    });
});

File: ruby/spec/thermostat_spec.rb
require 'thermostat'

describe Thermostat do

  describe "#temp" do
    it "has a default temperature of 20" do
      expect(subject.temperature).to eq Thermostat::DEFAULT_TEMPERATURE
    end
  end

  describe "#up" do
    it "increases the temperature" do
      subject.up
      expect(subject.temperature).to eq 21
    end

    
    it "has a maximum temperature of 25 when PSM ON" do
      6.times do
        subject.up
      end
      expect(subject.temperature).to eq 25
    end
    

    
    it "has a maximum temperature of 32 when PSM OFF" do
      subject.power_save_off
      13.times do
        subject.up
      end
      expect(subject.temperature).to eq 32
    end
    
  end

  describe "#down" do
    it "decreases the temperature" do
      subject.down
      expect(subject.temperature).to eq 19
    end

    it "has a minimum temperature of 10 degrees" do
      11.times { subject.down }
      expect(subject.temperature).to eq 10
    end
  end

  describe "#reset" do
    it "resets the temperature to 20" do
      subject.up
      subject.up
      subject.reset
      expect(subject.temperature).to eq Thermostat::DEFAULT_TEMPERATURE
    end
  end

  describe "in_psm?" do
    it "is in power save mode by default" do
      expect(subject).to be_in_psm
    end
  end

  describe "#power_save_off" do
    it "turns off power_save_mode" do
      subject.power_save_off
      expect(subject).not_to be_in_psm
    end
  end

  describe "#power_save_on" do
    it "turns on power_save_mode" do
      subject.power_save_off
      subject.power_save_on
      expect(subject).to be_in_psm
    end

    it "reduces the temperature to 25" do
      subject.power_save_off
      8.times { subject.up }
      expect(subject.temperature).to eq 28
      subject.power_save_on
      expect(subject.temperature).to eq 25
    end
  end

  describe "#energy_usage" do
   
    it "LOW when temp is below 18 degrees" do
      subject.down
      subject.down
      subject.down
      expect(subject.energy_usage).to eq :low
    end
    
    it "MEDIUM when temp is between 18 and 25 degrees" do
      expect(subject.energy_usage).to eq :medium
    end
  
    it "HIGH when temp is 25 or more degrees" do
      5.times { subject.up }
      expect(subject.energy_usage).to eq :high
    end
  end
  
end